### ISSUE TITLE
bulk update: exceeded maximum write batch size of 1000

### ISSUE THREAD

**Issue by cirias** [go-mgo/mgo#288]
When I use bulk to handle more than 1000 updates, I got the error `exceeded maximum write batch size of 1000`.

I found the reason is [Collection.writeOp](https://github.com/go-mgo/mgo/blob/v2/session.go#L4513) did not handle the 1000 limit for `bulkUpdateOp` and `bulkDeleteOp`.

This should be fix.

**Comment by pjebs** [comment]
MongoDb server only allows 1000 updates max

**Comment by cirias** [comment]
Yes. But according to https://docs.mongodb.com/manual/reference/limits/#Write-Command-Operation-Limit-Size the driver should not have the limit.

Also see https://github.com/go-mgo/mgo/issues/62

**Comment by pjebs** [comment]
Actually I agree with you (if you are correct about your reference).
That way, if mongodb decides to increase the max bulk limit, then the library doesn't need to be updated.

**Comment by epelc** [comment]
@cirias does this happen if your not using unordered mode(ie using ordered mode)?

Might be same issue as this https://jira.mongodb.org/browse/NODE-234

I also just ran into this but switching to using the default ordered mode fixed it.

**Comment by niemeyer** [comment]
Yes, this is a bug in the driver. We currently split inserts, and must split other operations too.

**Comment by asaf** [comment]
@niemeyer Any update on this? this blocks us and there's a PR waiting,

Anything that blocks the PR from being merged?

Thanks

**Comment by antoniomtz** [comment]
Waiting for this issue too.
The solution now is to split the Bulk into 1000? or will the max limit be removed?

**Comment by epelc** [comment]
@antoniomtz I suggest splitting at 500. We ran into errors splitting at 1000.

**Comment by antoniomtz** [comment]
@epelc thanks

**Comment by antoniomtz** [comment]
@epelc I have a question.
When I'm doing my split, bulk.Run() is taking the previous data that I added to my upsert.

I have a loop of:
```
bulk.Unordered()
bulk.Upsert(interface[range1:range2])
bulk.Run()
```
How do I reset my Bulk from previous data pass to Upsert?

I tried initializing Bulk in every split and get an error in mongo.

```
bulk:= session.db.collection.Bulk() // Getting error.
bulk.Unordered()
bulk.Upsert(interface[range1:range2])
bulk.Run()
```
Thanks.

**Comment by epelc** [comment]
@antoniomtz `Run runs all the operations queued up.`

ie you add operations to `bulk`, drain it by running them with `bulk.Run()` once you hit 500 ops, repeat

```
// psuedo code
mybulk := mgo.NewBulk()

mybulk.Unordered()

var ops int
for {
  if ops > 500 {
  // We need to run our operations before we queue up too many
  if err :=  bulk.Run(); err != nil {
    return err
  }
  }
  ops += 1
  bulk.Upsert(data)
}
// Do a final run if you have any ops left
if ops > 0 {
  if err :=  bulk.Run(); err != nil {
    return err
  }
}

// Your done
```

**Comment by antoniomtz** [comment]
@epelc HI, that is exactly my problem. Bulk.Run() doesn't seem to drain my data being queued up.
When I inspect Bulk, on the third bulk.Run() call, it has opcount:1500.

Any idea of what's happening?

Thanks

**Comment by epelc** [comment]
@antoniomtz Are you checking for errors? Post the code you are running or I can't help much.

**Comment by antoniomtz** [comment]
@epelc So, luckily, I found a github project that handles batches of unlimited size.

https://github.com/andreyvit/mongobulk/blob/master/bulk.go#L126 

This is what I was missing. You need to point bulk to nil and re-initialize it to "flush" any data.

Thanks

**Comment by danbogos** [comment]
Hey Guys,

We have also hit this issue. 
I was wondering if there is a consensus on the solution or should we each try fixing on our own?

Ta,
DanB

**Comment by mxplusb** [comment]
@danbogos I just updated the PR for this, so hopefully we can get some traction. I very much want this too...

**Comment by hexadecy** [comment]
@mxplusb it has been merged in this fork: https://github.com/globalsign/mgo
Fow now, probably the best fork to run with latest 3.2 or 3.4.

**Comment by qRoC** [comment]
In the master branch, the limit is now [increased](https://github.com/mongodb/mongo/blob/master/src/mongo/db/ops/write_ops.h#L61) to 100,000.

You can build from stable branch with the desired [value](https://github.com/mongodb/mongo/blob/v3.4/src/mongo/db/ops/write_ops_parsers.cpp#L47), because this is a stupid limit(look comment).
